src/main/java/org/springframework/integration/kafka/listener/MetadataStoreOffsetManager.java

```diff
@@ -0,0 +1,207 @@
+	 */
+	public void setReferenceTimestamp(long referenceTimestamp) {
+		this.referenceTimestamp = referenceTimestamp;
+	}
+
+	/**
+	 * @see OffsetManager#updateOffset(Partition, long)
+	 */
+	@Override
+	public void updateOffset(Partition partition, long offset) {
+		metadataStore.put(asKey(partition), Long.toString(offset));
+		offsets.put(partition, offset);
+	}
+
+	/**
+	 * @see OffsetManager#getOffset(Partition)
+	 */
+	@Override
+	public long getOffset(Partition partition) {
+		if(!offsets.containsKey(partition)) {
+			initializeOffsets(Collections.singleton(partition));
+		}
+		if (!offsets.containsKey(partition)) {
+			throw new IllegalStateException("Offset " + offsets + " cannot be found");
+		}
+		return offsets.get(partition);
+	}
+
+	@Override
+	public void resetOffsets(Collection<Partition> partitionsToReset) {
+		MutableMultimap<BrokerAddress, Partition> partitionsByLeaderAddress = groupBy(partitionsToReset, getLeaderFunction);
+		partitionsByLeaderAddress.toMap().forEachKeyValue(new RetrieveInitialOffsetsAndUpdateProcedure());
+	}
+
+	private void initializeOffsets(Collection<Partition> partitions) {
+		List<Partition> partitionsRequiringInitialOffsets = new ArrayList<Partition>();
+		for (Partition partition : partitions) {
+			if (!this.offsets.containsKey(partition)) {
+				String storedOffsetValueAsString = this.metadataStore.get(asKey(partition));
+				Long storedOffsetValue = null;
+				if (storedOffsetValueAsString != null) {
+					try {
+						storedOffsetValue = Long.parseLong(storedOffsetValueAsString);
+					}
+					catch (NumberFormatException e) {
+						LOG.warn("Invalid value: " + storedOffsetValue);
+					}
+				}
+				if (storedOffsetValue != null) {
+					offsets.put(partition, storedOffsetValue);
+				}
+				else {
+					partitionsRequiringInitialOffsets.add(partition);
+				}
+			}
+		}
+		if (partitionsRequiringInitialOffsets.size() > 0) {
+			resetOffsets(partitionsRequiringInitialOffsets);
+		}
+	}
+
+	private String asKey(Partition partition) {
+		return partition.getTopic() + " " + partition.getId() + " " + consumerId;
+	}
+
+	@SuppressWarnings("serial")
+	private class GetLeaderFunction implements Function<Partition, BrokerAddress> {
+		@Override
+		public BrokerAddress valueOf(Partition object) {
+			return connectionFactory.getLeader(object);
+		}
+	}
+
+	@SuppressWarnings("serial")
+	private class RetrieveInitialOffsetsAndUpdateProcedure implements Procedure2<BrokerAddress, RichIterable<Partition>> {
+		@Override
+		public void value(BrokerAddress brokerAddress, RichIterable<Partition> partitions) {
+			Partition[] partitionsAsArray = partitions.toArray(new Partition[partitions.size()]);
+			Result<Long> initialOffsets = connectionFactory.createConnection(brokerAddress).fetchInitialOffset(referenceTimestamp, partitionsAsArray);
+			if (initialOffsets.getErrors().size() == 0) {
+				for (Partition partition : partitions) {
+					offsets.put(partition, initialOffsets.getResults().get(partition));
```

Comment:
Yes, it is. The role of the  `metadataStore` would be store only offsets that the client has actually read, and therefore to be updated only through the `updateOffset` operation.